### PR TITLE
Support complete collection of set methods, including multi-argument forms

### PR DESCRIPTION
**Pull request by tetromino**
Also update the docs and tests.

Working towards https://github.com/google/starlark-go/issues/584

### CODE REVIEW

File: doc/spec.md [hunk 1]
```diff
@@ -3691,15 +3713,45 @@ x                                        # set([1, 3])
...
+### set·intersection\_update
+
+`S.intersection_update(*others)` removes from the set `S` any elements not found
+in at least one of the iterable sequences `*others`. It returns `None`.
+
+It is permissible to call `intersection_update` without any arguments; this
+leaves the set `S` unchanged.
+
+`intersection_update` fails if the set `S` is frozen or has active iterators, or
+if any element of any of the `*others` is unhashable.
```
Comment (adonovan) [inline doc/spec.md:3739]: For the record, this precludes a short cut when S is empty.

File: starlark/library.go [hunk 2]
```diff
@@ -2463,20 +2542,24 @@ func updateDict(dict *Dict, updates Tuple, kwargs []Tuple) error {
 	return nil
 }
 
-func setUpdate(s *Set, args Tuple, kwargs []Tuple) error {
+// forIterableArgsDo is a helper for implementing methods which perform the
+// same operation on an iteratover over each iterable argument, and don't
```
Comment (adonovan) [inline starlark/library.go:2546]: iterator

File: starlark/library.go [hunk 3]
```diff
@@ -2463,20 +2542,24 @@ func updateDict(dict *Dict, updates Tuple, kwargs []Tuple) error {
...
+// allow keyword arguments - for example, set.update(*others),
+// set.union(*others), etc.
+func forIterableArgsDo(operation func(other Iterator) error, args Tuple, kwargs []Tuple) error {
```
Comment (adonovan) [inline starlark/library.go:2549]: It's common style in Go to put the func parameter last so that it can be called like so:
```
forIterableArgsDo(args, kwarks, func() {
     ...
})
```

Of course, the error result makes this slightly less tidy, but I think it's still the preferred order.

File: starlark/value.go [hunk 4]
```diff
@@ -1221,20 +1221,18 @@ func (s *Set) clone() *Set {
...
-			return nil, err
-		}
+	if err := set.InsertAll(other); err != nil {
+		return nil, err
 	}
 	return set, nil
 }
 
-func (s *Set) InsertAll(iter Iterator) error {
+// TODO: rename to Update()
```
Comment (adonovan) [inline starlark/value.go:1232]: We can't rename it, but you can add Update, make this function a wrapper around it, and add a `//go:fix inline` annotation on InsertAll, which will nudge people to use Update.

File: starlark/value.go [hunk 5]
```diff
@@ -1275,7 +1294,10 @@ func (s *Set) IsSubset(other Iterator) (bool, error) {
 	}
 }
 
+// Warning: the resulting set has the iteration order of other.
```
Comment (adonovan) [inline starlark/value.go:1297]: Rather than add these Warnings and TODOs, we should fix the underlying problem in this CL.

File: starlark/value.go [hunk 6]
```diff
@@ -1295,17 +1317,33 @@ func (s *Set) Intersection(other Iterator) (Value, error) {
 
 func (s *Set) SymmetricDifference(other Iterator) (Value, error) {
 	diff := s.clone()
+	if err := diff.symmetricDifferenceUpdate(s, other); err != nil {
+		return nil, err
+	}
+	return diff, nil
+}
+
+// s and sReadonlyCopy must contain the same elements; s will be modified,
```
Comment (adonovan) [inline starlark/value.go:1326]: It's unfortunate that you need to clone s just to be able to iterate over it at the same time as (one of the callers) wishes to update s. Can `set_symmetric_difference_update` not simply call `s.symmetricDifferenceUpdate(s, iter)`? Is the awkwardness designed to handle the edge case `x.op_update(x)` for nonempty sets x? It might be simpler to shortcut in that case: x.symmetric_difference_update(x) becomes just x.clear().

We should add tests of such cases.

File: starlark/library.go [hunk 7]
```diff
@@ -2218,34 +2222,89 @@ func set_clear(_ *Thread, b *Builtin, args Tuple, kwargs []Tuple) (Value, error)
...
+		return nil, nameErr(b, err)
+	}
+	return None, nil
 }
 
 // https://github.com/google/starlark-go/blob/master/doc/spec.md#set_intersection.
 func set_intersection(_ *Thread, b *Builtin, args Tuple, kwargs []Tuple) (Value, error) {
-	// TODO: support multiple others: s.difference(*others)
+	accum := b.Receiver().(*Set).clone()
+	// TODO: use an intersection primitive that preserves lhs iteration order
```
Comment (adonovan) [inline starlark/library.go:2248]: We should address this in this CL. I think we should explicitly guarantee the insertion order of all these operations.

**Review by adonovan — COMMENTED** [review]
Thanks for contributing these operators.

File: starlark/value.go (hunk 6 above)
Comment (tetromino) [reply starlark/value.go:1326]: The problem is that `other` can be a list/tuple containing duplicates. Consider something like `set([x]).symmetric_difference([x, x])` - if we didn't have sReadonlyCopy, we would naively iterate over `[x, x]`, delete `x`, re-insert `x`, and end up with `set([x])` as the wrong answer.

So we have two possibilities: either make a temporary copy of the lhs set `s`, as I did here; or else make a temporary set copy of the rhs list `other` to deduplicate.

With the second approach, it makes more sense to pass `other` as a `Value` rather than an iterator, so that we can avoid making a copy if `other` is already deduplicated (e.g. already a set). And that approach also can work for order-preserving intersection. WDYT?

File: starlark/library.go (hunk 3 above)
Comment (tetromino) [reply starlark/library.go]: Thanks for the suggestion, makes the code more readable!

File: starlark/library.go (hunk 2 above)
Comment (tetromino) [reply starlark/library.go]: Fixed

File: starlark/value.go (hunk 6 above)
Comment (adonovan) [reply starlark/value.go:1326]: Ah, interesting example (and definitely worthy of a comment). Still I don't understand how the readonly copy of s solves that problem: we would need to first accumulate the iterator into a set s', and then compute sΔs', otherwise we would need to remember information (parity of occurrence) about all iterator values already seen.

> So we have two possibilities: either make a temporary copy of the lhs set s, as I did here; or else make a temporary set copy of the rhs list other to deduplicate.

The second approach seems conceptually simpler.

File: starlark/value.go (hunk 4 above)
Comment (tetromino) [reply starlark/value.go]: Done.

File: starlark/value.go (hunk 5 above)
Comment (tetromino) [reply starlark/value.go]: Done.

File: starlark/value.go (hunk 6 above)
Comment (tetromino) [reply starlark/value.go]: Done.

File: starlark/library.go (hunk 7 above)
Comment (tetromino) [reply starlark/library.go]: Done.

File: doc/spec.md (hunk 1 above)
Comment (tetromino) [reply doc/spec.md:3739]: That seems reasonable? We usually prohibit no-op mutation attempts - for example, if a set is immutable, we don't allow trying to update it with an element that it already contains.

**Comment by tetromino** [comment]
I've added a new Set.IntersectionUpdate() which preserves lhs iteration order - please take a look. Notably, since IntersectionUpdate needs to delete from the lhs hash table whilst iterating over it, it needs to iterate manually instead of using a normal iterator. I believe this is sound as long as the entry being deleted is an entry that has already been examined, and as long as we save the next pointer (since deletion will mutate the entry struct).
